src/lib/universal-client.js

```diff
@@ -0,0 +1,638 @@
+ * 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+ */
+
+/**
+ * The JavaScript library that implements the CometVisu protocol.
+ * 
+ * @module CometVisu Client
+ * @title CometVisu Client
+ * @reqires jQuery
+ */
+define(
+    [],
+    function() {
+      "use strict";
+
+      /**
+       * Class that handles the communicaton of the client *
+       * 
+       * @class CometVisu
+       * @constructor foo
+       * @param backend {String} name of the backend (cgi-bin|default|oh|openhab|oh2|openhab2)
+       * @param initPath {String|null} optional path to login ressource
+       */
+      function CometVisu(backend, initPath) {
+        this.initPath = initPath;
+
+        // used for backwards compability
+        this.aliases = {
+          'cgi-bin' : 'default',
+          'oh' : 'openhab',
+          'oh2' : 'openhab2'
+        };
+        if (this.aliases[backend]) {
+          backend = this.aliases[backend];
+        }
+
+        this.backends = {
+          'default' : {
+            name : 'default',
+            baseURL : '/cgi-bin/',
+            transport : 'long-polling',
+            resources : {
+              login : 'l',
+              read : 'r',
+              write : 'w',
+              rrd : 'rrdfetch'
+            },
+            hooks : {}
+          },
+          'openhab' : {
+            name : 'openHAB',
+            baseURL : '/services/cv/',
+            // keep the e.g. atmosphere tracking-id if there is one
+            resendHeaders : {
+              'X-Atmosphere-tracking-id' : undefined
+            },
+            // fixed headers that are send everytime
+            headers : {
+              'X-Atmosphere-Transport' : 'long-polling'
+            },
+            hooks : {
+              onClose : function() {
+                // send an close request to the openHAB server
+                var oldValue = this.headers["X-Atmosphere-Transport"];
+                this.headers["X-Atmosphere-Transport"] = "close";
+                $.ajax({
+                  url : this.config.baseURL + this.config.resources.read,
+                  dataType : 'json',
+                  context : this,
+                  beforeSend : this.beforeSend
+                });
+                if (oldValue != undefined) {
+                  this.headers["X-Atmosphere-Transport"] = oldValue;
+                } else {
+                  delete this.headers["X-Atmosphere-Transport"];
+                }
+              }
+            }
+          }
+        };
+        // init default settings
+        this.config = this.backends['default'];
+        if (backend && backend !== 'default') {
+          if ($.isPlainObject(backend)) {
+            // override default settings
+            $.extend(this.config, backend);
+          } else if (this.backends[backend]) {
+            // merge backend settings into this.config
+            $.extend(this.config, this.backends[backend]);
+          }
+        }
+
```

Comment:
Quite often in the code below a structure like `this.transport[this.config.transport]. ...` is used.

To help readability and performance (a very little) it could be a good idea to introduce a local variable here that contains `this.transport[this.config.transport]` and is used further down. (Due to the refcounting nature of JS the memory overhead is minimal)